Gems/Atom/RPI/Code/Source/RPI.Edit/Material/MaterialPropertyConnectionSerializer.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/RPI.Edit/Material/MaterialPropertyConnectionSerializer.h>
+#include <Atom/RPI.Edit/Material/MaterialUtils.h>
+
+#include <AzCore/Serialization/Json/BaseJsonSerializer.h>
+#include <AzCore/Serialization/Json/JsonSerializationResult.h>
+#include <AzCore/Serialization/Json/JsonSerialization.h>
+#include <AzCore/Serialization/Json/StackedString.h>
+
+namespace AZ
+{
+    namespace RPI
+    {
+        namespace JsonMaterialPropertyConnectionSerializerInternal
+        {
+            namespace Field
+            {
+                static constexpr const char type[] = "type";
+                static constexpr const char name[] = "name";
+                static constexpr const char id[] = "id"; // For backward compatibility
+                static constexpr const char shaderIndex[] = "shaderIndex";
+            }
+
+            static const AZStd::string_view AcceptedFields[] =
+            {
+                Field::type,
+                Field::name,
+                Field::id,
+                Field::shaderIndex
+            };
+        }
+
+        AZ_CLASS_ALLOCATOR_IMPL(JsonMaterialPropertyConnectionSerializer, SystemAllocator, 0);
+
+        JsonSerializationResult::Result JsonMaterialPropertyConnectionSerializer::Load(void* outputValue, const Uuid& outputValueTypeId,
+            const rapidjson::Value& inputValue, JsonDeserializerContext& context)
+        {
+            namespace JSR = JsonSerializationResult;
+            using namespace JsonMaterialPropertyConnectionSerializerInternal;
+
+            AZ_Assert(azrtti_typeid<MaterialTypeSourceData::PropertyConnection>() == outputValueTypeId,
+                "Unable to deserialize material property connection to json because the provided type is %s",
+                outputValueTypeId.ToString<AZStd::string>().c_str());
+            AZ_UNUSED(outputValueTypeId);
+
+            MaterialTypeSourceData::PropertyConnection* propertyConnection = reinterpret_cast<MaterialTypeSourceData::PropertyConnection*>(outputValue);
+            AZ_Assert(propertyConnection, "Output value for JsonMaterialPropertyConnectionSerializer can't be null.");
+
+            JSR::ResultCode result(JSR::Tasks::ReadField);
+
+            if (!inputValue.IsObject())
+            {
+                return context.Report(JsonSerializationResult::Tasks::ReadField, JsonSerializationResult::Outcomes::Unsupported, "Property connection must be a JSON object.");
+            }
+            
+            MaterialUtils::CheckForUnrecognizedJsonFields(AcceptedFields, AZ_ARRAY_SIZE(AcceptedFields), inputValue, context, result);
```

Comment:
An alternative way to do this that won't require the explicit field names is to take the number of expected fields and substract every `ContinueLoadingFromJsonObjectField` that returns that defaults are used, then compare that value to the number of values in the JSON Object. If they don't match than there are unused fields.